### PR TITLE
handling grafana url for k8Version 1.15 and below

### PR DESCRIPTION
**Pull request by rashmirai21**
# Description

Fixes # (https://github.com/devtron-labs/devtron/issues/316)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Following graphs have been test for k8s version k8sVersion 1.15.0, below and above

K8sVersion below 1.15.0
Aggregate (App Details and Modal)
- [x] CPU 
- [x] RAM
- [x] Throughput
- [x] Latency

Per Pod (App Details and Modal)
- [x] CPU 
- [x] RAM
- [x] Throughput
- [x] Latency

K8sVersion Above 1.15.0
Aggregate (App Details and Modal)
- [x] CPU 
- [x] RAM
- [x] Throughput
- [x] Latency

Per Pod (App Details and Modal)
- [x] CPU 
- [x] RAM
- [x] Throughput
- [x] Latency

# Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/app/details/appDetails/AppMetrics.tsx [hunk 1]
```diff
@@ -18,7 +18,7 @@ import { URLS } from '../../../../config';
 import { getHostURLConfiguration } from '../../../../services/service';
 
 
-export const AppMetrics: React.FC<{ appName: string, environment, podMap: Map<string, any> }> = ({ appName, environment, podMap }) => {
+export const AppMetrics: React.FC<{ k8sVersion, appName: string, environment, podMap: Map<string, any> }> = ({ k8sVersion, appName, environment, podMap }) => {
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/AppMetrics.tsx:21]: whenever we are adding new attribute we should add it at the end

File: src/components/app/details/appDetails/GraphsModal.tsx [hunk 2]
```diff
@@ -86,15 +87,15 @@ export class GraphModal extends Component<GraphModalProps, GraphModalState>{
...
-        let status4xx = getIframeSrc(this.props.appId, this.props.envId, this.props.environmentName, 'status', this.props.newPodHash, this.state.calendarInputs, tab, false, '4xx');
-        let status5xx = getIframeSrc(this.props.appId, this.props.envId, this.props.environmentName, 'status', this.props.newPodHash, this.state.calendarInputs, tab, false, '5xx');
-        let status = getIframeSrc(this.props.appId, this.props.envId, this.props.environmentName, 'status', this.props.newPodHash, this.state.calendarInputs, tab, false, 'Throughput');
-        let throughput = getIframeSrc(this.props.appId, this.props.envId, this.props.environmentName, 'status', this.props.newPodHash, this.state.calendarInputs, tab, false, 'Throughput');
-        let mainChartUrl = getIframeSrc(this.props.appId, this.props.envId, this.props.environmentName, this.state.mainChartName, this.props.newPodHash, this.state.calendarInputs, tab, true, this.state.statusCode);
+        let cpu = getIframeSrc(this.props.appId, this.props.envId, this.props.environmentName, 'cpu', this.props.newPodHash, this.state.calendarInputs, tab, false, this.props.k8sVersion);
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/GraphsModal.tsx:90]: When we have so many arguments, its bettter to encapsulate it in 1 or 2 objects

File: src/components/app/details/appDetails/utils.tsx [hunk 3]
```diff
@@ -213,27 +213,56 @@ export function getCalendarValue(startDateStr: string, endDateStr: string): stri
     return str;
 }
 
-export function getIframeSrc(appId: string | number, envId: string | number, environmentName: string, chartName: ChartTypes, newPodHash: string, calendarInputs, tab: AppMetricsTabType, isLegendRequired: boolean, statusCode?: string):string {
+export function isK8sVersion115AndBelow(k8sVersion: string): boolean {
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/utils.tsx:216]: isK8sVersion115OrBelow

File: src/components/app/details/appDetails/utils.tsx [hunk 4]
```diff
@@ -213,27 +213,56 @@ export function getCalendarValue(startDateStr: string, endDateStr: string): stri
...
+    let target = [1, 15, 0];
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/utils.tsx:217]: should be [1,15,1000]

File: src/components/app/details/appDetails/utils.tsx [hunk 5]
```diff
@@ -213,27 +213,56 @@ export function getCalendarValue(startDateStr: string, endDateStr: string): stri
...
+    let version: string[] = []
+    try {
+        version = (k8sVersion.split("v")[1]).split(".");
+    } catch (error) {
+
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/utils.tsx:222]: what are we doing with error? if we have error then we dont know... so we should make an assumption and return true or false. There is no reason to move forward

File: src/components/app/details/appDetails/utils.tsx [hunk 6]
```diff
@@ -213,27 +213,56 @@ export function getCalendarValue(startDateStr: string, endDateStr: string): stri
...
-export function getIframeSrc(appId: string | number, envId: string | number, environmentName: string, chartName: ChartTypes, newPodHash: string, calendarInputs, tab: AppMetricsTabType, isLegendRequired: boolean, statusCode?: string):string {
+export function isK8sVersion115AndBelow(k8sVersion: string): boolean {
+    let target = [1, 15, 0];
+    let version: string[] = []
+    try {
+        version = (k8sVersion.split("v")[1]).split(".");
+    } catch (error) {
+        
+    }
+    let versionNum: number[] = version.map(item => Number(item));
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/utils.tsx:224]: What if this line gives error?

File: src/components/app/details/appDetails/utils.tsx [hunk 7]
```diff
@@ -213,27 +213,56 @@ export function getCalendarValue(startDateStr: string, endDateStr: string): stri
...
+        }
     }
     switch (chartName) {
         case 'latency':
             url += `latency/latency`;
             break;
         case 'ram':
-            url += `memory/memory-usage`;
+            if (isK8sVersion115AndBelow(k8sVersion)) {
+                url = `${url}/memory-usage-k8s15`;
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/utils.tsx:257]: in this line we have `/` in line 259, we dont have slash, also lets use standard mechanism to concatenate url  either like this or like line 259

File: src/components/app/details/appDetails/GraphsModal.tsx [hunk 8]
```diff
@@ -86,15 +88,29 @@ export class GraphModal extends Component<GraphModalProps, GraphModalState>{
...
+        }
+        let appInfo = {
+            appId: this.props.appId,
+            envId: this.props.envId,
+            environmentName: this.props.environmentName,
+            newPodHash: this.props.newPodHash,
+            k8sVersion: k8sVersion,
+        }
+
+        let cpu = getIframeSrc(appInfo, 'cpu', this.state.calendarInputs, tab, false);
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/GraphsModal.tsx:104]: generally we should extract out constants like 'cpu', 'ram', 'latency' at the top of the file

File: src/components/app/details/appDetails/GraphsModal.tsx [hunk 9]
```diff
@@ -86,15 +88,29 @@ export class GraphModal extends Component<GraphModalProps, GraphModalState>{
...
+        let ram = getIframeSrc(appInfo, 'ram', this.state.calendarInputs, tab, false);
+        let latency = getIframeSrc(appInfo, 'latency', this.state.calendarInputs, tab, false);
+        let status2xx = getIframeSrc(appInfo, 'status', this.state.calendarInputs, tab, false, '2xx');
+        let status4xx = getIframeSrc(appInfo, 'status', this.state.calendarInputs, tab, false, '4xx');
+        let status5xx = getIframeSrc(appInfo, 'status', this.state.calendarInputs, tab, false, '5xx');
+        let status = getIframeSrc(appInfo, 'status', this.state.calendarInputs, tab, false, 'Throughput');
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/GraphsModal.tsx:110]: Throughput has T capital, should be avoided, either all small case or all capital

File: src/components/app/details/appDetails/GraphsModal.tsx [hunk 10]
```diff
@@ -159,7 +175,20 @@ export class GraphModal extends Component<GraphModalProps, GraphModalState>{
     }
 
     handleChartChange(chartName: ChartTypes, status?: string): void {
-        let mainChartUrl = getIframeSrc(this.props.appId, this.props.envId, this.props.environmentName, chartName, this.props.newPodHash, this.state.calendarInputs, this.state.tab, true, this.state.statusCode);
+        let k8sVersion = this.props.k8sVersion;
+        let defaultK8sVersion = 'v16.1.10';
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/GraphsModal.tsx:179]: use v16.1.0

File: src/components/app/details/appDetails/utils.tsx [hunk 11]
```diff
@@ -213,27 +214,82 @@ export function getCalendarValue(startDateStr: string, endDateStr: string): stri
...
-    else {
-        url = `${rootUrl}/grafana/d-solo/NnFpQOKGk/res_status_per_pod`;
+    return true;
+}
+
+export function isK8sVersion115OrBelow(k8sVersion: string): boolean {
+    let target = [1, 15, 1000];
+    let version: string[] = [];
+    let versionNum: number[] = [];
+    version = (k8sVersion.split("v")[1]).split(".");
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/utils.tsx:236]: duplicate code, same as isK8sVersionValid, we should extract duplicate code

File: src/components/app/details/appDetails/utils.tsx [hunk 12]
```diff
@@ -213,27 +214,82 @@ export function getCalendarValue(startDateStr: string, endDateStr: string): stri
...
+    return grafanaURL;
+}
+
+export function getGrafanaBaseURL(k8sVersion, chartName): string {
+    let rootUrl = process.env.REACT_APP_ORCHESTRATOR_ROOT.replace('/orchestrator', '');
+    let url = '';
+    if (isK8sVersion115OrBelow(k8sVersion) && (chartName === 'cpu' || chartName === 'ram')) {
+        url = `${rootUrl}/grafana/d/devtron-app-metrics-`;
+    } else {
+        if (chartName !== 'status') {
```
Comment (pghildiyal) [inline src/components/app/details/appDetails/utils.tsx:268]: is this same as (chartName === 'cpu' || chartName === 'ram') ? if yes we should use same condition at both the places also we can simplify by setting

let url = `${rootUrl}/grafana/d-solo/NnFpQOKGk/res_status_per_pod`;
if (chartName === 'cpu' || chartName === 'ram') {
   if (isK8sVersion115OrBelow(k8sVersion)) {
            url = `${rootUrl}/grafana/d/devtron-app-metrics-`;
    } else {
            url = `${rootUrl}/grafana/d-solo/devtron-app-metrics-`;
    }
}

File: src/components/app/details/appDetails/GraphsModal.tsx (hunk 9 above)
Comment (rashmirai21) [reply src/components/app/details/appDetails/GraphsModal.tsx:109]: Throughput is also a query parameter, where `T` is required to be capital. I have extracted it out in a constant.

File: src/components/app/details/appDetails/utils.tsx (hunk 12 above)
Comment (rashmirai21) [reply src/components/app/details/appDetails/utils.tsx]: There is a change in url. Following is the current logic:

let url = '/grafana/d-solo';
if (chartName === 'status') {
   url = `${url}/NnFpQOKGk/res_status_per_pod`;
 }
 else {
     url = `${url}/devtron-app-metrics-`;
}
return url;

**Review by pghildiyal — APPROVED** [review]

**Review by pghildiyal — APPROVED** [review]

**Review by nishant-d — APPROVED** [review]

**Pull request merged**